[ENHANCEMENT] Patrol Frequency

scribblecrumb opened the pull request:
<!-- Write BELOW The Headers and ABOVE The comments else it may not be viewable. -->
<!-- You can view CONTRIBUTING.md for a detailed description of the pull request process. -->
<!-- Be sure to name your PR something descriptive and succinct; include Bugfix: Feature: Enhancement: or Content: in the title to describe what type of PR it is. -->
<!-- IF YOU ARE DOING A BUGFIX: Please target the latest release branch if the bug that you are fixing is also present in the latest release. -->

## About The Pull Request
I've implemented the Frequency and auto-weighting system from short events into patrols.

- `weight` has been replaced by `frequency` in the patrol format (documentation and schemas are updated to account for this)
- This does mean that every patrol file appears altered. I utilized a script to convert `weight` to `frequency`, thus all the patrol files were also reformatted (which is nice! some of the files had pretty mangled indents). However, I don't expect reviewers to read through all these patrols. ^^; Please trust that I've calibrated the new frequencies to fit close to the original intentions for the events.
- Each patrol event and patrol outcome now determines it's own `weight` according to the ways it's been constrained. 
- Added `get_frequency` and `find_new_frequency` helper funcs to `event_filters.py` so that frequency calcs could be shared by patrols and short events in a centralized space.
- Added print lines for frequency/weight of patrol and outcome to the patrol print message
 
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage senior developers from merging your PR! -->

## Why This Is Good For ClanGen
This helps unbury highly constrained patrols from the pile and make them more likely to occur when their constraints are being met.  The frequency system is also far easier for writers to understand and gives clear direction instead of just unbridled freedom. Freedom is great, but it becomes a hindrance when working with such a large range of events and attempting to establish consistency between a large team.
<!-- If this is a bug fix, you can remove this section. -->
<!-- Please add a short description of why you think these changes would benefit the game. -->
<!-- If this PR was the result of discussion/testing on the discord, please add a link to the discord conversation here. -->
<!-- If you have multiple features that can stand on their own, or unrelated bugfixes, please create separate PRs for them. -->

## Proof of Testing
<img width="377" height="100" alt="image" src="https://github.com/user-attachments/assets/d984fb45-7ff4-415a-8d72-21a5dae83304" />
<img width="559" height="105" alt="image" src="https://github.com/user-attachments/assets/264c1026-a3e1-4b76-92ca-607858765cd0" />
<img width="535" height="111" alt="image" src="https://github.com/user-attachments/assets/5aedd4bc-34f2-4743-946f-e76fe1e07544" />
<img width="460" height="69" alt="image" src="https://github.com/user-attachments/assets/c8e44747-a13d-4d44-b7f9-9e9e1ee8e258" />
<img width="561" height="104" alt="image" src="https://github.com/user-attachments/assets/d990994c-1e08-42ac-a7f2-4470e73d5124" />
<img width="565" height="101" alt="image" src="https://github.com/user-attachments/assets/31f16027-b864-4755-8aec-7c2e66890de0" />
<img width="554" height="103" alt="image" src="https://github.com/user-attachments/assets/3b6498e9-2b4d-46de-8ff6-2981358f3164" />


I've also got the beta testers playing on my branch to give feedback on the frequency of events: [thread](https://discord.com/channels/1003759225522110524/1455331612299628680)
Been playing a bit with it myself and adjusting things as I see fit, I think though that it's hit a decent balance for now!
<!-- Include any screenshots, debugging steps, or links to beta testing threads here. At least one form of proof of testing is REQUIRED for all new content. You must be able to run the code locally before you PR it here. -->

## Changelog/Credits
- Patrols now utilize the same frequency and weight system as short events
- Documentation updated to reflect this
- Some plains patrols got misplaced into the forest files, they've been put back where they belong
- Hopefully you will see less badgers
<!-- Include any changes that should be made to the changelog of the game here or any changes to the credits file of the game. -->
<!-- This is just for easy access later for senior developers gathering this information; this process is not automated. -->

scribblecrumb committed:
move frequency calc to centralized func

scribblecrumb committed:
remove outdated documentation

scribblecrumb committed:
reformat outcomes to frequency

scribblecrumb committed:
adjust schema to outcome frequency

scribblecrumb committed:
reformat patrols to frequency

scribblecrumb committed:
Update patrol.schema.json

scribblecrumb committed:
implementing weight and frequency across patrols

scribblecrumb committed:
update print message

scribblecrumb committed:
assessing frequency assignments

scribblecrumb committed:
adjust hunting weight balancing

scribblecrumb committed:
beach hunting frequency assessment

scribblecrumb committed:
med initial reassesments

scribblecrumb committed:
training and med frequency

scribblecrumb committed:
move plains patrols back where they belong

scribblecrumb committed:
forest hunting

scribblecrumb committed:
forest training

scribblecrumb committed:
general patrols

scribblecrumb committed:
mountain border

scribblecrumb committed:
mountain hunting

scribblecrumb committed:
mountain training

scribblecrumb committed:
plains hunting

scribblecrumb committed:
plains training

scribblecrumb committed:
find new freq helper

scribblecrumb committed:
fixing while loop

scribblecrumb committed:
json fixes

scribblecrumb committed:
badger balance

scribblecrumb committed:
new cat and other clan balancing

scribblecrumb committed:
small adjustments

scribblecrumb committed:
buff rel constraint weight boost

scribblecrumb committed:
lowering 'no one steps up' frequency

scribblecrumb committed:
fix: crash when only romantic patrols found

scribblecrumb committed:
fix: incorrect tagging

scribblecrumb committed:
documentation

scribblecrumb committed:
frequency adjustments

scribblecrumb committed:
fix endless loop

scribblecrumb committed:
frequency adjustments

scribblecrumb committed:
Merge remote-tracking branch 'upstream/development' into patrol-frequency

scribblecrumb committed:
Merge remote-tracking branch 'upstream/development' into patrol-frequency

scribblecrumb committed:
merge fixes

scribblecrumb committed:
merge fixes

scribblecrumb committed:
merge fixes

scribblecrumb committed:
merge fix

scribblecrumb committed:
Update any.json

scribblecrumb committed:
Update any.json

scribblecrumb committed:
Merge remote-tracking branch 'upstream/development' into patrol-frequency

scribblecrumb committed:
fix merge

j-gynn commented on scripts/events_module/patrol/patrol_outcome.py:
```diff
@@ -43,12 +44,21 @@
+
+    num_of_ages = len(CatAge)
+
+    num_of_ranks = CatRank.get_num_of_clan_ranks()
```
since these are constants, shouldn't they be in UPPER_SNAKE_CASE?

Unfortunately we can't have these in `constants.py` I think because of `Personality`, since iirc that pulls from `Cat` because of course it does.

scribblecrumb replied:
Ah I guess I wasn't thinking about them as constants, but they definitely are in retrospect lmao I'll change that

j-gynn reviewed (commented):
i hope you don't mind but i have not read a single patrol change <3 only the code. another constants question from me, but once you're happy with the balance from betas, I'm happy to approve.

scribblecrumb committed:
constant name style changes

also remove age and rank consts from patroloutcome since they weren't actually used

scribblecrumb commented:
Not reading the patrol jsons is exactly what I expected and encourage lmao

Honestly what I've heard from the betas sounds good to me, so I'm happy to get this merged and then fine-tune things as the community provides more feedback via #3896. All in all it already sounds to be a good improvement on the prior system.

j-gynn reviewed (approved):
lgtm
